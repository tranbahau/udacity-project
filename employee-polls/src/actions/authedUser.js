const SET_AUTHED_USER_ID = 'SET_AUTHED_USER_ID';

const LOG_OUT = 'LOG_OUT';

export const setAuthedUser = (userId) => {
  return {
    type: SET_AUTHED_USER_ID,
    userId,
  };
};

export const setUserLogOut = () => {
  return {
    type: LOG_OUT,
  };
};
