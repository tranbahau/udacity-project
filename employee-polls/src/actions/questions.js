const GET_QUESTIONS = 'GET_QUESTIONS';
const SAVE_QUESTION_ANSWER = 'SAVE_QUESTION_ANSWER';
const SAVE_QUESTION = 'SAVE_QUESTION';

export const getQuestions = (questions) => {
  return {
    type: GET_QUESTIONS,
    questions,
  };
};

export const saveQuestionAnswer = (answerInfo) => {
  return {
    type: SAVE_QUESTION_ANSWER,
    answerInfo,
  };
};

export const saveQuestion = (question) => {
  return {
    type: SAVE_QUESTION,
    question,
  };
};
