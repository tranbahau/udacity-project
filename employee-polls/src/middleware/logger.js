export const logger = (store) => (next) => (action) => {
  console.group(action.type);
  console.log(action);
  console.groupEnd();
  next(action);
};
