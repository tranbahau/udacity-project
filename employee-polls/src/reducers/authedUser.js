export const authedUser = (state = null, action) => {
  switch (action.type) {
    case 'LOG_OUT':
      return null;
    case 'SET_AUTHED_USER_ID':
      return action.userId;
    default:
      return state;
  }
};
