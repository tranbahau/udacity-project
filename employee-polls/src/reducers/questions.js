export const questions = (state = {}, action) => {
  switch (action.type) {
    case 'GET_QUESTIONS':
      return { ...state, ...action.questions };
    case 'SAVE_QUESTION_ANSWER':
      return {
        ...state,
        [action.answerInfo.questionId]: {
          ...state[action.answerInfo.questionId],
          [action.answerInfo.answer]: {
            ...state[action.answerInfo.questionId][action.answerInfo.answer],
            votes: state[action.answerInfo.questionId][
              action.answerInfo.answer
            ].votes.concat(action.answerInfo.authedUser),
          },
        },
      };
    case 'SAVE_QUESTION':
      return {
        ...state,
        [action.question.id]: action.question,
      };
    default:
      return state;
  }
};
