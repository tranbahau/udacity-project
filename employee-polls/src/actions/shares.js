import { showLoading, hideLoading } from 'react-redux-loading-bar';
import { _getQuestions, _getUsers, _saveQuestionAnswer } from '../utils/_DATA';
import { getUsers } from './users';
import { getQuestions, saveQuestion, saveQuestionAnswer } from './questions';
import { _saveQuestion } from '../utils/_DATA';
import { setAuthedUser } from './authedUser';

export const handleGetUsers = () => {
  return (dispatch) => {
    dispatch(showLoading());

    return _getUsers()
      .then((users) => dispatch(getUsers(users)))
      .then(() => dispatch(hideLoading()));
  };
};

export const handleGetQuestions = () => {
  return (dispatch) => {
    dispatch(showLoading());

    return _getQuestions()
      .then((questions) => dispatch(getQuestions(questions)))
      .then(() => dispatch(hideLoading()));
  };
};

export const handleSaveQuestionAnswer = (
  { questionId, authedUser, answer },
  callBack
) => {
  return (dispatch) => {
    return _saveQuestionAnswer({
      qid: questionId,
      authedUser: authedUser,
      answer: answer,
    })
      .then(() =>
        dispatch(
          saveQuestionAnswer({
            questionId,
            authedUser,
            answer,
          })
        )
      )
      .then(() =>
        callBack(`/questions/${questionId}`, {
          answer,
          authedUser,
          questionId,
        })
      );
  };
};

export const handleSaveQuestion = (
  { optionOneText, optionTwoText, author },
  callBack
) => {
  return (dispatch) => {
    return _saveQuestion({
      optionOneText,
      optionTwoText,
      author,
    })
      .then((formattedQuestion) => dispatch(saveQuestion(formattedQuestion)))
      .then(() => callBack('/'));
  };
};

export const handleLogin = (userId) => {
  return (dispatch) => {
    dispatch(setAuthedUser(userId));
  };
};
