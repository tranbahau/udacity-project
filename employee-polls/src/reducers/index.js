import { combineReducers } from 'redux';
import { users } from './users';
import { questions } from './questions';
import { authedUser } from './authedUser';
import { loadingBarReducer } from 'react-redux-loading-bar';

export const reducers = combineReducers({
  users: users,
  questions: questions,
  authedUser: authedUser,
  loadingBar: loadingBarReducer,
});
