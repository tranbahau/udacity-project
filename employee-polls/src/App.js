import { Fragment, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Route, Routes } from 'react-router-dom';
import { handleGetQuestions, handleGetUsers } from './actions/shares';
import './App.css';

import QuestionType from './components/Questions/Question/QuestionType/QuestionType';
import NewQuestion from './components/NewQuestion/NewQuesiton';
import Leaderboard from './components/Leaderboard/Leaderboard';
import Questions from './components/Questions/Questions';
import NavContainer from './components/NavContainer/NavContainer';
import Login from './components/Login/Login';
import ProtectedRoute from './routes/ProtectedRoute';
import NotFound from './components/NotFound/NotFound';
import LoadingBar from 'react-redux-loading-bar';

function App() {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(handleGetUsers());
    dispatch(handleGetQuestions());
    // eslint-disable-next-line
  }, []);

  return (
    <Fragment>
      <LoadingBar />
      <NavContainer />
      <Routes>
        <Route path="/login" element={<Login />} />
        <Route
          path="/"
          element={
            <ProtectedRoute>
              <Questions />
            </ProtectedRoute>
          }
        />
        <Route
          path="/questions/:id"
          element={
            <ProtectedRoute>
              <QuestionType />
            </ProtectedRoute>
          }
        />
        <Route
          path="/leaderboard"
          element={
            <ProtectedRoute>
              <Leaderboard />
            </ProtectedRoute>
          }
        />
        <Route
          path="/add"
          element={
            <ProtectedRoute>
              <NewQuestion />
            </ProtectedRoute>
          }
        />
        <Route path="*" element={<NotFound />} />
      </Routes>
    </Fragment>
  );
}

export default App;
