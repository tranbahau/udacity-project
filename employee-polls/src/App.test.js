import { reducers } from './reducers';
import { Provider } from 'react-redux';

import App from './App';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import thunk from 'redux-thunk';
import { logger } from './middleware/logger';
import { setAuthedUser } from './actions/authedUser';
import { configureStore } from '@reduxjs/toolkit';

describe('App', () => {
  const store = configureStore({
    reducer: reducers,
    middleware: [thunk, logger],
  });
  it('should render the component', () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it('should show NavBar page when logged in', () => {
    store.dispatch(setAuthedUser('tylermcginnis'));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    const heading = component.getByTestId('nav-container');
    expect(heading).toBeInTheDocument();
  });
});
