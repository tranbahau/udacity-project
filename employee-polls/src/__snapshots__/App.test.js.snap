// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <nav
        class="container navbar navbar-expand-lg navbar-light bg-light"
        data-testid="nav-container"
      >
        <div
          class="container"
        >
          <span
            class="navbar-brand"
            to="/"
          >
            Employee Polls Project
          </span>
          <button
            aria-controls="basic-navbar-nav"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="basic-navbar-nav"
          >
            <div
              class="me-auto navbar-nav"
            >
              <a
                class="navlink"
                href="/"
              >
                Questions
              </a>
              <a
                class="navlink"
                href="/leaderboard"
              >
                Leaderboard
              </a>
              <a
                class="navlink"
                href="/add"
              >
                New
              </a>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="container"
      >
        <div
          class="loginForm"
        >
          <p
            class="title"
          >
            Employee Polls
          </p>
          <p
            class="text"
          >
            Login
          </p>
          <form>
            <select
              class="form-select"
              style="width: 280px;"
            >
              <option
                value="default"
              >
                Please choose user to login
              </option>
            </select>
            <button
              class="btnSubmit"
              disabled=""
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="direction: ltr;"
    >
      <div
        class=""
        style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
      />
      <div
        style="display: table; clear: both;"
      />
    </div>
    <nav
      class="container navbar navbar-expand-lg navbar-light bg-light"
      data-testid="nav-container"
    >
      <div
        class="container"
      >
        <span
          class="navbar-brand"
          to="/"
        >
          Employee Polls Project
        </span>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              class="navlink"
              href="/"
            >
              Questions
            </a>
            <a
              class="navlink"
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              class="navlink"
              href="/add"
            >
              New
            </a>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="loginForm"
      >
        <p
          class="title"
        >
          Employee Polls
        </p>
        <p
          class="text"
        >
          Login
        </p>
        <form>
          <select
            class="form-select"
            style="width: 280px;"
          >
            <option
              value="default"
            >
              Please choose user to login
            </option>
          </select>
          <button
            class="btnSubmit"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
